### PR TITLE
Merge bitcoin#17708: prevector: avoid misaligned member accesses

### PR DESCRIPTION
**Pull request by pravblockc**
Backport: Merge #17708: prevector: avoid misaligned member accesses
add ubsan supressions for below errors
runtime error: load of value 210, which is not a valid value for type 'bool'
runtime error: null pointer passed as argument 1, which is declared to never be null !=

### CODE REVIEW

**Review by UdjinM6 — DISMISSED** [review]
LGTM, utACK

**Comment by PastaPastaPasta** [comment]
Why is https://github.com/dashpay/dash/pull/4451/commits/54bbfb443197ca8414f76aeb16fedbc3e95ee974 here? Seems to just revert part of the merge?

**Comment by pravblockc** [comment]
> Why is [54bbfb4](https://github.com/dashpay/dash/commit/54bbfb443197ca8414f76aeb16fedbc3e95ee974) here? Seems to just revert part of the merge?

No, Actually I have added to suppress below warning which are Dash specific. It has to be resolved at a later stage along with others. In case if is not acceptable, would you prefer to resolve now itself?
runtime error: load of value 210, which is not a valid value for type 'bool'
runtime error: null pointer passed as argument 1, which is declared to never be null !=

**Comment by PastaPastaPasta** [comment]
Why should we suppress these warnings though? They are proper warnings right that should be fixed?

File: src/prevector.h [hunk 1]
```diff
@@ -565,7 +565,8 @@ class prevector {
             if (v.size() != s) {
                 v.resize(s);
             }
-            ::memmove(v.data(), &*b, s);
+            if (v.size() != 0)
```
Comment (PastaPastaPasta) [inline src/prevector.h:568]: can we use !v.empty() here instead?

Comment (pravblockc) [reply src/prevector.h:568]: agree

**Comment by pravblockc** [comment]
As we discussed, dash specific errors/warnings has been fixed without suppression.
@UdjinM6 @PastaPastaPasta

File: src/prevector.h [hunk 2]
```diff
@@ -565,7 +565,8 @@ class prevector {
             if (v.size() != s) {
                 v.resize(s);
             }
-            ::memmove(v.data(), &*b, s);
+            if (!v.empty())
+                ::memmove(v.data(), &*b, s);
```
Comment (PastaPastaPasta) [inline src/prevector.h:569]: brackets or same line please

Comment (pravblockc) [reply src/prevector.h]: done

**Review by UdjinM6 — APPROVED** [review]
utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merging via merge commit

**Pull request merged**